Steinhardt class: A new single class for the Steinhardt order Parameter

b-butler opened the pull request:
## Description
The goal is to create a class Steinhardt that better emulates the functionality
of other classes within freud.order with the Steinhardt order parameter. This
functionality is currently encapsulated into the LocalQl and LocalWl classes.
This new class would in theory become the new standard API will leaving the
other classes for backwards-compatibility.

Some goals for this branch are ...
+ Provide a single compute method in the python interface
+ Use flags at construction to determine the particular order parameter given by
  compute.
+ Keep current and improve further - when applicable - performance of
  class with an eye towards potenial further parallelization.
+ Improve cpp code readability and better emulate the DRY mantra.

## Motivation and Context
Most other classes within the order module have a single compute method exposed
to the python interface. However, with the Steinhardt order parameters currently
8 exist between 2 classes. Furthermore, the C++ code use to determine the
Steinhardt order parameter requires code repetition. The current python
interface also exposes the user to this complexity. Until recently, the logic
for the compute methods were confusingly split between the Cython and C++
layers. Further improvements to code readability and maintainability are
warrented and this new API would allow for that.

## Pull Request Status
- [x] Create draft Steinhardt.cc. This should initially mostly be copying code
  from LocalQl and LocalWl to file.
- [x] Create working compute method.
  - [x] Impliment logic using member flags to calculate only what is necessary.
  - [x] Abstract array reallocation to function that only reallocates used
    arrays.
  - [x] Potentially move individual computational steps in compute method to
    private functions.
  - [ ] Add parallelization to loops over particle number.
- [x] Clean up code where applicable.
- [x] Add Steinhardt class to \_order.pxd.
- [x] Add Steinhardt class to order.pyx.
- [x] Create tests for Steinhardt class.
- [x] Update documentation

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

b-butler committed:
Merge remote-tracking branch 'origin/parallelize_steinhardt' into parallelize_steinhardt

b-butler committed:
Adding conditional array resizing for computeAve.

b-butler committed:
Moving Ql compute from Cython to cpp for AveQl.

b-butler committed:
Adding some comments.

b-butler committed:
Preventing unnecessary array reallocation.

b-butler committed:
Creating Steinhardt class header file.

b-butler committed:
Updated header file and added initial body.

Adding function for reallocating arrays.
Moved compute base_compute.
Begin sructuring of compute method logic in c++.

b-butler committed:
Beginning implimentation of compute method.

Also removing unnecessary bits from the different compute methods.

b-butler committed:
1st complete implementation of Steinhardt class.

This commit was done when the c++ code compiled.

b-butler committed:
Edits that were required to compile.

Some intermediate changes came from the parallize_Steinhardt branch.
These changes were finished since they should improve speed. However,
they could be reasonably checkouted from another branch to just limit
the branch to the Steinhardt class. The changes involve moving more
logic to the c++ side rather than Cython.

b-butler committed:
Added Cython wrapper for Steinhardt class.

Compiles with --ENABLE-CYTHON flag on install. Some logic behind getting
the correct order parameter had to be on the Cython side since Ql and Wl
are of different types. Therefore, I defined a Python function that
selectively calls the right Cython wrapped getter on the c++ side to be
used by the Python API.

b-butler committed:
Adding tests for Steinhardt class.

I just copied and tweaked existing unit test code. Therefore, there may
be bugs present or the checks may not be completely suited to the new
class.

b-butler committed:
Reduced conditionals on constructor.

Removed redundent conditionals and colapsed the assignment of
orderParameter pointers in one conditional grouping.

b-butler committed:
Merge remote-tracking branch 'origin/master' into Steinhardt_class

b-butler committed:
Changing Local{Q,W}l classes back to master.

Previous changes broke tests, so I just moved them back to master.

b-butler committed:
Changing implimentation of getting order parameter

Previous method would have taken too much time. Logic to return correct
order parameter was moved to the getter in c++. Logic in Cython did not
have to change.

b-butler committed:
Modified tests for Steinhardt class.

b-butler committed:
Reseting Cython for Local{Q,W}l classes.

b-butler committed:
Changing test class name to match class.

b-butler committed:
Working on implimenting Wl correctly.

Ql tests correctly already, so as soon as Wl works, the branch can be
merged to master.

b-butler committed:
Working unit tests and class functioning.

Steinhardt branch works the same as Local{W,Q}l now. However, one brief
caevot is that I had to turn over the WRITTABLE option for the LocalQl
tests to pass, and I did not impliment that on the Steinhardt unit
tests.

b-butler committed:
Adding documentation for Steinhardt class.

b-butler committed:
Adding const to memoryviews to fix Cython bug.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,495 @@
+// Copyright (c) 2010-2019 The Regents of the University of Michigan
```
Can you retab this file? It looks like in a lot of places you end up with 8 spaces instead of 4 for tabs.

vyasr replied:
Also make sure you follow our (admittedly strange) code style with respect to brace indentation, see [here](https://freud.readthedocs.io/en/latest/sourceconventions.html#cpp)

b-butler replied:
Will do

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,495 @@
+    parallel_for(tbb::blocked_range<size_t>(0, m_Np),
+        [=] (const blocked_range<size_t>& r)
+        {
+	// May not be necessary check
```
What check isn't necessary? Is there something we can remove here, or is this a copied comment from old code?

b-butler replied:
Copied from old code

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,495 @@
+        const float rmaxsq = m_rmax * m_rmax;
+        const float normalizationfactor = 4*M_PI/(2*m_l+1);
+        const size_t *neighbor_list(nlist->getNeighbors());
+	// --------------------------
```
Remove

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,495 @@
+                const vec3<float> delta = m_box.wrap(points[j] - ref);
+                const float rsq = dot(delta, delta);
+
+		// Replace conditional with function call above.
```
Is this your comment?

b-butler replied:
No that I think was left over from the hackathon, but I must have missed it

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -0,0 +1,230 @@
+            }
+
+		//! Get the last calculated order parameter
+		std::shared_ptr<float> getQl()
```
Fix indentation

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -0,0 +1,230 @@
+
+
+
+	private:
```
Unindent `private`

vyasr replied:
Also remove a couple of the extra blank lines above

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -0,0 +1,230 @@
+		void reallocateArrays(unsigned int Np);
+
+		//! Calculates the base Ql order parameter before further modifications
+		//if any.
```
Add a space `// if any`

vyasr commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+    original definition by the average value of :math:`\overline{Q}_{lm}(k)`
+    over all the :math:`k` neighbors of particle :math:`i` as well as itself.
+
+    The :data:`norm` constructor arguement provides
```
type `argument`

vyasr commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+    particles.
+
+    .. moduleauthor:: Xiyu Du <xiyudu@umich.edu>
+    .. moduleauthor:: Vyas Ramasubramani <vramasub@umich.edu>
```
Add yourself! :)

vyasr commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+        rmin (float):
+            Can look at only the second shell or some arbitrary RDF region.
+        average (bool):
+            Determines whether to calcuate the averaged Steinhardt order
```
typo `calculate`

vyasr commented on tests/test_order_Steinhardt.py:
```diff
@@ -0,0 +1,90 @@
+        positions = np.random.uniform(-box.Lx/2, box.Lx/2,
+                                      size=(N, 3)).astype(np.float32)
+
+        positions.flags['WRITEABLE'] = False
```
For consistency with our other tests, I would remove this here and in other tests whether it's still present.

vyasr reviewed (changes requested):
There are some style issues, but the content all looks fine to me. There are a few places where it would be nice to combine a little more of the logic, but unfortunately different code paths differ just enough that it's not possible.

bdice commented on cpp/order/Steinhardt.h:
```diff
@@ -0,0 +1,230 @@
+        virtual void computeYlm(const float theta, const float phi,
+                                std::vector<std::complex<float> > &Ylm);
+
+		//! Reallocates only the necesary arrays when the number of particles changes
```
typo `necessary`

bdice commented on freud/_order.pxd:
```diff
@@ -74,6 +74,20 @@ cdef extern from "TransOrderParameter.h" namespace "freud::order":
+cdef extern from "Steinhardt.h" namespace "freud::order":
+    cdef cppclass Steinhardt:
+        Steinhardt(const freud._box.Box &, float, unsigned int, float,
+                bool, bool, bool) except +
```
You'll need to fix the code here to address the `flake8` code style issues that are preventing tests from running. See here for info about the `flake8` pre-commit hook, which can help you catch these early: https://freud.readthedocs.io/en/stable/sourceconventions.html#python

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+    R"""Compute the local Steinhardt [Steinhardt1983]_rotationally invariant
+    :math:`Q_l` :math:`Wl` order parameter for a set of points.
+
+    Implements the local rotationally invariant :math:`Q_l` or :math:`Wl` order
```
`W_l` needs an underscore

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+
+    The :data:`norm` constructor arguement provides
+    normalized versions of the plain :math:`Ql` or :math:`Wl` or normalized
+    average if the :data:`average` flag is set to true.  where the normalization
```
Fix the incomplete sentence here. Maybe write "Normalization is performed by..."

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+            The last computed selected variant of the Steinhardt order
+            parameter for each particle (filled with NaN for particles with no
+            neighbors).
+    .. todo:: move box to compute, this is old API
```
We'll want to actually make this change - the `box` shouldn't be in the constructor anymore, only in the `compute` method.

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+        return np
+
+    @property
+    def St(self):
```
I discussed with @vyasr a while back - I think we'd like to use the property name `order` for all order parameters' data going forward. Since this is a new class, let's rename `St` to `order`.

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+
+    @property
+    def St(self):
+        if self.stptr.getUseWl():
```
I'd like to see this logic implemented on the C++ level.

vyasr replied:
To clarify, you want there to be a single function on the C++ side that returns the correct order parameter depending on whether or not the flag indicates to use Wl or Ql? Keep in mind that the two currently return different types of output, so I'm not sure you can accomplish that.

I also don't think the Wl really needs to be complex from what I recall of the math defining it, but IIRC you do get meaningful imaginary components from the Wl classes. I haven't really used them in years, so I don't know what to make of that, and I would suggest holding off making any changes to the types, etc, for a separate PR where we do more explicit validation checking against e.g. pyboo.

b-butler replied:
I would have liked to impliment the logic in C++, but vyasr is right. The type difference between Wl and Ql prevents the logic from being on the C++ side.

bdice replied:
Just saw this - editing my last comment about next steps since this isn't possible. Sorry for the confusion.

bdice commented on freud/order.pyx:
```diff
@@ -494,6 +494,173 @@ cdef class TransOrderParameter:
+    cdef freud.box.Box m_box
+    cdef rmax
+
+    def __cinit__(self, box, rmax, l, rmin=0, average=False, norm=False,
```
We'll probably refactor this to use the new neighbor-finding code after #272 has been merged. Then we will handle the neighbor-finding logic (e.g. `rmax`, `rmin`, `k` nearest neighbors) via a `QueryArgs` argument to the `compute` method. That way, we'll also encompass the behavior of the current `LocalQlNear` and `LocalWlNear` classes.

vyasr replied:
Agreed. For now it's fine that this only supports the distance cutoffs.

bdice commented:
@b-butler Thanks for your work on this! It will be a really nice feature improvement. We'll do a few iterations on this to get it right, since this feature is so heavily used. Ping @vyasr and I if you have questions or when you've made a round of changes and are ready for more review.

b-butler committed:
Fixing style and adding suggestions.

* Fixing the C++ and Cython style
* Removing box from constructor and to compute method
* Updating and fixing documentation
* Removing some comments

b-butler committed:
Merge remote-tracking branch 'origin/master' into Steinhardt_class

b-butler committed:
Fixed bugs in new compute method and test.

* Tests for Steinhardt class work again.
* Box object is only required for compute method not init.

b-butler committed:
Adding changes to impliment repr unittest.

Just added the necessary changes to impliment a repr unittest to improve
test coverage.

b-butler committed:
Fixing some documentation errors.

b-butler committed:
Moving class function to private.

Moving the reduce method for parallelizing the base_compute method to
private, since it shouldn't be called outside the class.

b-butler commented:
I addressed all the comments and fixed some other stuff I noticed.

bdice committed:
Cleanup and minor edits.

bdice committed:
Merge remote-tracking branch 'origin/master' into Steinhardt_class

bdice commented:
@b-butler Good work so far! A few broad tasks for refactoring and simplification:

1. Many of the computations (e.g. normalization) are repeated through the code. For example, I would prefer to see one "normalize" call that accepts shared pointers as arguments and uses the flags of the class to determine how to specialize, instead of `computeNorm`, `computeAveNorm`, `computeNormWl`, `computeAveNormWl`.
2. If the `norm` flag is false, do we need to compute `m_Qlm` at all? From what I can tell, this data is only used if `norm` is true. It would be good to verify that and make that step conditional if possible.
3. There is a loop in `computeAve` that is labeled "Can be parallelized." It would be good to parallelize this.

bdice committed:
Fix includes.

bdice committed:
Update convert_array.

bdice committed:
Template array allocations and rename AveQlm[i] to Qlm[i]Ave.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,510 @@
+    for (unsigned int i = 0; i < m_Np; i++)
+        {
+        // Revert Ql normalization
+        m_Qli.get()[i] /= normalizationfactor;
```
It doesn't appear like this result is being used - I looked at the original `LocalWl` to confirm. The original code had an option for `enableNormalization` that causes this to be useful, but it doesn't appear that it was ever possible to control this on the Python side. I will remove this in the current changes I am making to the code.

bdice replied:
^ done.

bdice committed:
Specify test precision, use assert_allclose.

bdice committed:
Add aggregateWl method to reduce repeated code.

bdice commented:
I did some clean up and reduced the code repetition for the `Wl` methods. There is still more to be done with reducing repetition, but it may not be exactly what I was originally saying above about `norm` methods. Suggestions 2 and 3 are still valid. Here's the task list I have in mind:
- [ ] Only compute Qlm if needed
- [ ] Parallelize `computeAve` loop
- [ ] Refactor remaining blocks of repeated code into functions
- [ ] Remove unnecessary member variable arrays (e.g. it doesn't make sense to have multiple member variables for `m_WliAveNorm` ... `m_Wli`, etc. if only one of them can be used at a time).

I'm thinking about creating a class that wraps around our shared pointer arrays and plays nicely with Cython, taking the `makeArray` method I just wrote for this PR to the next level. We might combine some nice behaviors of shared pointers and `std::vector` to create a wrapper layer that handles things like using `memset` to zero out the data, easily instantiating an array with a default deleter, making it simpler to iterate over the data, etc. This kind of code is repeated lots of places in freud and it's a prime target for refactoring in my opinion just because it would help simplify our codebase a lot.  Thoughts, @vyasr @jinsoo960?

bdice committed:
Refactor Wl computations, divide m_Qlm and m_QlmAve by number of particles at compute time.

bdice committed:
Refactor normalization logic.

bdice commented:
It's strange to me that we return full arrays for the "normalized" order parameters - it's exactly the same value for every particle.

bdice committed:
Merge branch 'master' into Steinhardt_class

bdice commented:
I finished the refactoring tasks I wanted to do - @b-butler, please take a look and feel free to pick up with the remaining tasks when you have time.

b-butler commented:
Will do @bdice. Just saw all your commits and recommendations.

b-butler committed:
Start moving norm to scalar and Ave parallization

* Move norm calculation to scalar
* Pallelize ComputeAve loop

b-butler committed:
Merge branch 'origin/master' into Steinhardt_class

b-butler committed:
Adding norm to Cython interface.

* Adding norm property to class
* Removing norm flag from constructor
* changing _ql property to Ql to expose property to interface

b-butler committed:
Completing introduction of norm attribtute.

* Added tests for working norm.
* Fixed memory bug in class.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+// Calculating Ylm using fsph module
+void Steinhardt::computeYlm(const float theta,
+						    const float phi,
+							std::vector<std::complex<float> > &Ylm)
```
Fix spacing here.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+    unsigned int j(0);
+    // old definition in compute (theta: 0...pi, phi: 0...2pi)
+    // in fsph, the definition is flipped
```
I'd delete this comment, it's referring to a pre-fsph implementation of Steinhardt that must be 3 years old at this point.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+void Steinhardt::compute(const box::Box& box,
+						 const locality::NeighborList *nlist,
+						 const vec3<float> *points,
+						 unsigned int Np)
```
Fix spacing.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+    if (m_Np != Np)
+        {
+        m_Np = Np;
+        Steinhardt::reallocateArrays(Np);
```
You shouldn't need this qualification, just `reallocateArrays` is enough.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+    if (m_average)
+        {
+        Steinhardt::computeAve(box, nlist, points);
```
Same as above.

vyasr replied:
I see this in a number of places, I'd just drop all unnecessary `Steinhardt::`.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+    if (m_Wl)
+        {
+        memset((void*) m_WliOrder.get(), 0, sizeof(complex<float>)*m_Np);
```
Shouldn't this already have been zeroed by `reallocateArrays`?

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+        if (m_average)
+            {
+            Steinhardt::aggregateWl(m_WliOrder, m_QlmiAve);
+			m_NormWl = Steinhardt::normalizeWl();
```
Fix spacing.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+        else
+            {
+            Steinhardt::aggregateWl(m_WliOrder, m_Qlmi);
+			m_NormWl = Steinhardt::normalizeWl();
```
Fix spacing.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+        else
+            {
+            m_Norm = Steinhardt::normalize();
+            }
```
I'm confused as to why this conditional is here.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+void Steinhardt::baseCompute(const box::Box& box,
+							 const locality::NeighborList *nlist,
+							 const vec3<float> *points)
```
Fix spacing.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+    nlist->validate(m_Np, m_Np);
+
+    memset((void*) m_Qlmi.get(), 0, sizeof(complex<float>)*(2*m_l+1)*m_Np);
+    memset((void*) m_Qli.get(), 0, sizeof(float)*m_Np);
```
These should be set by `reallocateArrays`.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+				m_Qlm_local.local() = new complex<float> [2*m_l+1];
+				memset((void*) m_Qlm_local.local(), 0, sizeof(complex<float>) * (2*m_l+1));
+				}
+			}
```
Fix the indentation of this block, and I think you can also get rid of the memset since `reallocateArrays` does it.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+					if (!m_average)
+						{
+						m_Qlm_local.local()[k] += m_Qlmi.get()[index] / float(m_Np);
+						}
```
Fix indentation.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+void Steinhardt::computeAve(const box::Box& box,
+							const locality::NeighborList *nlist,
+							const vec3<float> *points)
```
Indentation

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+    const size_t *neighbor_list(nlist->getNeighbors());
+
+    memset((void*) m_QlmiAve.get(), 0, sizeof(complex<float>)*(2*m_l+1)*m_Np);
+    memset((void*) m_QliAve.get(), 0, sizeof(float)*m_Np);
```
Dealt with in reallocateArrays

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,391 @@
+
+void Steinhardt::reduce()
+	{
+	memset((void*) m_Qlm.get(), 0, sizeof(complex<float>)*(2*m_l+1));
```
Remove in favor of reallocateArrays

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -0,0 +1,201 @@
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <tbb/tbb.h>
```
I think you can remove some of these (cstring, stdexcept).

vyasr reviewed (changes requested):
This is nice, much shorter than the older methods and will significantly clean up that whole chunk of code. I still have some change requests, but it's getting there!

The formatting is still all over the place. Since you first started working on this PR, we've started using clang-format as our C++ style requirement. Can you merge in master and then try to apply that clang-format to your code (let Bradley or me know if you need help doing that)? That should help fix a lot of the spacing issues.

b-butler committed:
Changing the reallocateArrays method.

* Also some format fixes.

b-butler committed:
Merge origin/master into Steinhardt_class

b-butler committed:
Clang formatted Steinhardt class.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -30,151 +28,154 @@ void Steinhardt::computeYlm(const float theta,
-            m_Norm = Steinhardt::normalize();
-            }
+        {
+            m_Norm = normalize();
```
Seems like this `if` is redundant?

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -30,151 +28,154 @@ void Steinhardt::computeYlm(const float theta,
                 const vec3<float> delta = box.wrap(points[j] - ref);
                 const float rsq = dot(delta, delta);
 
                 if (rsq < rmaxsq && rsq > rminsq)
```
See if tests still pass if you remove this `if` statement. We have a problem with this check, see #297. We may need to change the preprocessing we do to the NeighborList to correct the behavior after removing this line.

bdice replied:
Specifically we need to check the behavior where `rmin > 0`, when there are neighbors within that inner shell distance.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,208 +185,196 @@ void Steinhardt::baseCompute(const box::Box& box,
+                }
+
+                const vec3<float> rn = points[n];
+                // rin = rn - ri, from i pointing to j.
```
Fix comment, `from i pointing to neighbor n`.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,208 +185,196 @@ void Steinhardt::baseCompute(const box::Box& box,
+                const vec3<float> rin = box.wrap(rn - ri);
+                const float rinsq = dot(rin, rin);
+
+                if (rinsq < rmaxsq && rinsq > rminsq)
```
See above comment about removing this check.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -30,151 +28,154 @@ void Steinhardt::computeYlm(const float theta,
-			}
+        const float normalizationfactor = 4 * M_PI / (2 * m_l + 1);
+        const size_t* neighbor_list(nlist->getNeighbors());
+
```
Add comment: `// Initialize thread-local m_Qlm and compute it in this function, if we won't average over neighbors later`

bdice replied:
Add a similar comment below in the average method, something like `// Initialize thread-local m_Qlm for storing neighbor-averaged QlmiAve, reduced over particles`

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,208 +185,196 @@ void Steinhardt::baseCompute(const box::Box& box,
-		}); //End parallel function
-    }
+    parallel_for(tbb::blocked_range<size_t>(0, m_Np), [=](const blocked_range<size_t>& r) {
+        bool QlmAve_exists;
```
Suggesting a change from `QlmAve_exists` to `Qlm_exists` as above, so it's clear these represent the same thing.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -0,0 +1,380 @@
+        if (m_average)
+        {
+            aggregateWl(m_WliOrder, m_QlmiAve);
+            m_NormWl = normalizeWl();
```
Move this below the `if` statement (so it's only specified once).

bdice reviewed (changes requested):
This is looking great. A few comments left. I'm really happy with how this is turning out, it's much simpler. 🙌

b-butler committed:
Adding requested c++ changes to Steinhardt class.

b-butler committed:
Adding num_neighbors specification for Steinhardt

* optional argument for constructor now lets the the number of neighbors
to calculate the order parameter over is now available.

b-butler committed:
Fixing c++ syntax styling using clang-format.

b-butler committed:
Adding Cython generated cpp files.

b-butler committed:
Merge branch 'master' into Steinhardt_class

b-butler commented:
The tests are currently failing in the github tests, but work in my local environment.

vyasr committed:
Merge remote-tracking branch 'origin/master' into Steinhardt_class

vyasr committed:
Fix flake8 and usage of convert_array. Also add common.cpp to repo

vyasr commented:
@b-butler the failing tests weren't really your fault. You needed to merge up to current master and make a couple minor changes to your calls to `convert_array`, but also there were issues with the `common` module being moved from pure Python to Cython that were causing problems. Should be fixed now. Remind me to review in the next day or so if I haven't yet.

b-butler commented:
@vyasr thanks for figuring it out!

b-butler committed:
Removing unnecessary commment.

vyasr committed:
Update Steinhardt.h

Fix spacing.

vyasr reviewed (approved):
@bdice I'm happy with this PR and I think it's ready. If you're happy with this as well, I recommend merging #317 and then making the change here to the enumerable_thread_specific array we use here, then merging in this PR.

b-butler committed:
Merge branch 'Steinhardt_class' of https://github.com/glotzerlab/freud into Steinhardt_class

vyasr committed:
Update CPP

vyasr committed:
Merge remote-tracking branch 'origin/master' into Steinhardt_class
